src/main/java/com/conveyal/datatools/manager/controllers/api/DeploymentController.java

```diff
@@ -84,6 +88,46 @@ private static Deployment deleteDeployment (Request req, Response res) {
         return deployment;
     }
 
+    /**
+     * HTTP endpoint for downloading a build artifact (e.g., otp build log or Graph.obj) from S3.
+     */
+    private static String downloadBuildArtifact (Request req, Response res) {
+        Deployment deployment = checkDeploymentPermissions(req, res);
+        DeployJob.DeploySummary summaryToDownload = null;
+        String uriString = null;
+        // If a jobId query param is provided, find the matching job summary.
+        String jobId = req.queryParams("jobId");
+        if (jobId != null) {
+            for (DeployJob.DeploySummary summary : deployment.deployJobSummaries) {
+                if (summary.jobId.equals(jobId)) {
+                    summaryToDownload = summary;
+                    break;
+                }
+            }
+        } else {
+            summaryToDownload = deployment.latest();
+        }
+        if (summaryToDownload == null) {
+            // Try to construct the URI string
+            OtpServer server = Persistence.servers.getById(deployment.deployedTo);
+            if (server == null) {
+                uriString = String.format("s3://%s/bundles/%s/%s/%s", "S3_BUCKET", deployment.projectId, deployment.id, jobId);
+                logMessageAndHalt(req, 400, "Cannot construct URI for build artifact. " + uriString);
```

Comment:
Seems like this error message could be better worded to indicate that the deployment doesn't have server information.